sycl/doc/extensions/supported/sycl_ext_intel_esimd/examples/histogram_2d.md

```diff
@@ -0,0 +1,186 @@
+  int i;
+  for (i = 0; i < width * height; i++) {
+    cpuHistogram[srcY[i]] += 1;
+  }
+}
+
+void writeHist(unsigned int *hist) {
+  int total = 0;
+
+  std::cerr << " Histogram: \n";
+  for (int i = 0; i < NUM_BINS; i += 8) {
+    std::cerr << "\n  [" << i << " - " << i + 7 << "]:";
+    for (int j = 0; j < 8; j++) {
+      std::cerr << "\t" << hist[i + j];
+      total += hist[i + j];
+    }
+  }
+  std::cerr << "\nTotal = " << total << " \n";
+}
+
+int checkHistogram(unsigned int *refHistogram, unsigned int *hist) {
+
+  for (int i = 0; i < NUM_BINS; i++) {
+    if (refHistogram[i] != hist[i]) {
+      return 0;
+    }
+  }
+  return 1;
+}
+
+int main(int argc, char *argv[]) {
+  unsigned int width = IMG_WIDTH * sizeof(unsigned int);
+  unsigned int height = IMG_HEIGHT;
+
+  // ------------------------------------------------------------------------
+
+  // Allocate Input Buffer
+  queue q;
+
+  auto dev = q.get_device();
+  unsigned char *srcY = malloc_shared<unsigned char>(width * height, q);
+  if (srcY == NULL) {
+    std::cerr << "Out of memory\n";
+    exit(1);
+  }
+
+  unsigned int *bins = malloc_shared<unsigned int>(NUM_BINS, q);
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+  uint range_width = width / BLOCK_WIDTH;
+  uint range_height = height / BLOCK_HEIGHT;
+
+  // Initializes input.
+  unsigned int input_size = width * height;
+  std::cerr << "Processing inputs\n";
+
+  srand(2009);
+  for (int i = 0; i < input_size; ++i) {
+    srcY[i] = rand() % 256;
+  }
+
+  for (int i = 0; i < NUM_BINS; i++) {
+    bins[i] = 0;
+  }
+
+  // ------------------------------------------------------------------------
+  // CPU Execution:
+
+  unsigned int cpuHistogram[NUM_BINS];
+  memset(cpuHistogram, 0, sizeof(cpuHistogram));
+  histogram_CPU(width, height, srcY, cpuHistogram);
+
+  try {
+    // create ranges
+    // We need that many workitems
+    auto GlobalRange = range<2>(range_width, range_height);
+    // Number of workitems in a workgroup
+    auto LocalRange = range<2>(1, 1);
+    nd_range<2> Range(GlobalRange, LocalRange);
+
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Hist>(
+          Range, [=](nd_item<2> ndi) SYCL_ESIMD_KERNEL {
+            using namespace sycl::ext::intel::esimd;
+
+            // Get thread origin offsets
+            uint h_pos = ndi.get_group(0) * BLOCK_WIDTH;
+            uint v_pos = ndi.get_group(1) * BLOCK_HEIGHT;
+
+            // Declare a 1xBLOCK_WIDTH uchar matrix to store the input block
+            // pixel value
+            simd<unsigned char, BLOCK_WIDTH> in;
+
+            // Declare a vector to store the local histogram
+            simd<unsigned int, NUM_BINS> histogram(0);
+
+            in.copy_from(srcY + v_pos * width + h_pos);
+
+        // Accumulate local histogram for each pixel value
+#pragma unroll
+            for (int j = 0; j < 32; j++) {
+              histogram.select<1, 1>(in[j]) += 1;
```

Comment:
1) This syntax is odd. Much more clear is to write:
```suggestion
              histogram[in[j]] += 1;
```
2) Do you see a better/more-efficient way doing this 32-iteration scalar loop processing it element-by-element?